### PR TITLE
Document ways to make services wait for dependencies

### PR DESCRIPTION
**Pull request by aanand**
- Moved the section to the top of the FAQ, on the assumption that it'll help people landing from searches
- Reword the question in the hope that it'll catch people searching for "startup order" or similar 
- Replace the paragraphs describing health checks with example implementations
- Also updated the next question while I was there, now that we support `stop_signal`

### CODE REVIEW

File: docs/faq.md [hunk 1]
```diff
@@ -40,6 +111,12 @@ in your Dockerfile.
 * If you are able, modify the application that you're running to
 add an explicit signal handler for `SIGTERM`.
 
+* Set the `stop_signal` to a signal which the application knows how to handle:
+
+      web:
+        build: .
+        stop_signal: SIGTERM
```
Comment (dnephin) [inline docs/faq.md]: `SIGTERM` is already the default, so I think this should be `SIGKILL` or `SIGINT`?

Comment (aanand) [inline docs/faq.md]: Oh whoops, I meant to put SIGINT yes

File: docs/faq.md [hunk 2]
```diff
@@ -15,7 +15,78 @@ weight=90
...
+To handle this, your application should attempt to re-establish a connection to
+the database after a failure. If the application retries the connection,
+it should eventually be able to connect to the database.
+
+There are several ways to implement this:
+
+-   Use a tool such as [wait-for-it](https://github.com/vishnubob/wait-for-it)
+    or [dockerize](https://github.com/jwilder/dockerize). These are small
+    wrapper scripts which you can include in your application's image and will
+    wait for TCP to be available on a given host and port.
```
Comment (dnephin) [inline docs/faq.md]: I think this should be "a TCP socket" or something like that. Feels like there is a missing word here. Maybe even just "wait for service to be available".

File: docs/faq.md [hunk 3]
```diff
@@ -15,7 +15,78 @@ weight=90
...
+The problem of waiting for a database to be ready is really just a subset of a
+much larger problem of distributed systems. In production, your database could
+become unavailable or move hosts at any time. Your application needs to be
+resilient to these types of failures.
+
+To handle this, your application should attempt to re-establish a connection to
+the database after a failure. If the application retries the connection,
+it should eventually be able to connect to the database.
+
+There are several ways to implement this:
```
Comment (dnephin) [inline docs/faq.md]: Technically these examples are not of application level retries, but workarounds for applications that don't properly retry.

**Comment by dnephin** [comment]
LGTM when green

File: docs/faq.md [hunk 4]
```diff
@@ -15,7 +15,77 @@ weight=90
...
+            build: .
+            ports:
+              - "80:8000"
+            depends_on:
+              - "db"
+            entrypoint: ./wait-for-it.sh db:5432
+          db:
+            image: postgres
+
+-   Write yor own wrapper script to perform a more application-specific health
```
Comment (simonvanderveldt) [inline docs/faq.md]: Minor typo here "yor"

**Comment by simonvanderveldt** [comment]
@aanand Nice to see this addressed!

File: docs/faq.md (hunk 4 above)
Comment (aanand) [inline docs/faq.md]: Thanks! Fixed it.

**Comment by dnephin** [comment]
LGTM

**Comment by bfirsh** [comment]
NICE! Missed this. :sparkling_heart:

**Pull request merged**